### PR TITLE
SONARJAVA-4770 S2438 FN on arguments whose concrete type is Thread

### PR DESCRIPTION
**Pull request by kaufco**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/ThreadAsRunnableArgumentCheck.java [hunk 1]
```diff
@@ -19,85 +19,131 @@
...
+      case RETURN_STATEMENT -> visitReturnStatement((ReturnStatementTree) tree);
+      case YIELD_STATEMENT -> visitYieldStatement((YieldStatementTree) tree);
+      case NEW_ARRAY -> visitNewArray((NewArrayTree) tree);
+      case ASSIGNMENT -> {
+        var assignment = (AssignmentExpressionTree) tree;
+        checkTypeCoercion(assignment.variable().symbolType(), assignment.expression());
+      }
+      case METHOD_INVOCATION -> {
+        var invacation = (MethodInvocationTree) tree;
+        visitInvocation(invacation.methodSymbol(), invacation.arguments());
```
Comment (johann-beleites-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/ThreadAsRunnableArgumentCheck.java:69]:
```suggestion
        var invocation = (MethodInvocationTree) tree;
        visitInvocation(invocation.methodSymbol(), invocation.arguments());
```

File: java-checks/src/main/java/org/sonar/java/checks/ThreadAsRunnableArgumentCheck.java [hunk 2]
```diff
@@ -19,85 +19,131 @@
...
+      }
+      case NEW_CLASS -> {
+        var invacation = (NewClassTree) tree;
+        visitInvocation(invacation.methodSymbol(), invacation.arguments());
+      }
+      default -> throw new IllegalArgumentException();
```
Comment (johann-beleites-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/ThreadAsRunnableArgumentCheck.java:75]: IMO, this is dead code that we won't be able to test and we shouldn't add this line. I know we have a rule (S131 I believe) that will complain, but I don't think it really makes sense in contexts like this and would consider it a FP. Not really convinced that it is useful enough as defensive programming here either. I'm going to leave the final decision up to you though.

**Review by johann-beleites-sonarsource — COMMENTED** [review]
Looks nice, a few cosmetic comments

File: java-checks/src/main/java/org/sonar/java/checks/ThreadAsRunnableArgumentCheck.java (hunk 2 above)
Comment (kaufco) [reply java-checks/src/main/java/org/sonar/java/checks/ThreadAsRunnableArgumentCheck.java]: Removed it. Now violating Joshua Bloch 6:36. May the lord forgive me.

File: java-checks/src/main/java/org/sonar/java/checks/ThreadAsRunnableArgumentCheck.java [hunk 3]
```diff
@@ -19,85 +19,131 @@
...
     }
-    return parametersTypes.get(index);
   }
 
-  private static String getMessage(ExpressionTree argument, Type providedType, int index) {
-    String array = providedType.isArray() ? "[]" : "";
-    return MessageFormat.format("Replace \"{0}\" of type Thread{1} with an instance of Runnable{1}.", getArgName(argument, index), array);
+  private void visitYieldStatement(YieldStatementTree tree) {
+    Tree enclosing = ExpressionUtils.getEnclosingTree(tree, Tree.Kind.SWITCH_EXPRESSION, Tree.Kind.SWITCH_STATEMENT);
+    if (enclosing == null || enclosing.is(Tree.Kind.SWITCH_STATEMENT)) {
```
[the remark this answers is not in the document]
Comment (kaufco) [reply java-checks/src/main/java/org/sonar/java/checks/ThreadAsRunnableArgumentCheck.java:129]: Added a comment about it.

**Review by johann-beleites-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**